### PR TITLE
Streamline preparing job across jobs; adding pre and post run hooks

### PR DESCRIPTION
**Pull request by samuelbroscheit**
Was triggered by https://github.com/uma-pi1/kge/pull/124 's request to add a pre and post run hook. Turned out that there was code duplication for preparing jobs. Now every job has a pre and post run hook. And every job gets its _prepare called *once* from the Job-level.

### CODE REVIEW

File: kge/job/job.py [hunk 1]
```diff
@@ -131,7 +141,29 @@ def _load(self, checkpoint: Dict):
...
+        if not self._is_prepared:
+            self._prepare()
+
+        for f in self.pre_run_hooks:
+            f(self)
+
+        result = self._run()
+
+        for f in self.post_run_hooks:
+            f(self)
```
Comment (rgemulla) [inline kge/job/job.py:162]: pass result here

File: kge/job/train.py [hunk 2]
```diff
@@ -790,9 +783,7 @@ def __init__(self, config, dataset, parent_job=None, model=None):
 
     def _prepare(self):
         """Construct dataloader"""
-
-        if self.is_prepared:
-            return
+        super()._prepare()
```
Comment (rgemulla) [inline kge/job/train.py:786]: Is this consistently added everywhere?

**Comment by rgemulla** [comment]
Ready from my point of view.

**Comment by samuelbroscheit** [comment]
You deleted this merge?

**Comment by samuelbroscheit** [comment]
> You deleted this merge?

Sorry got confused by githubs new interface ... :(

**Pull request merged**